// snack-string/index.js
exports.isEmpty = require('./lib/isEmpty');
exports.isBlank = require('./lib/isBlank');
exports.contains = require('./lib/contains');
exports.startsWith = require('./lib/startsWith');
exports.endsWith = require('./lib/endsWith');

exports.before = require('./lib/before');
exports.beforeLast = require('./lib/beforeLast');
exports.after = require('./lib/after');
exports.afterLast = require('./lib/afterLast');
exports.between = require('./lib/between');

exports.pad = require('./lib/pad');
exports.padLeft = require('./lib/padLeft');
exports.padRight = require('./lib/padRight');

exports.addStart = require('./lib/addStart');
exports.addEnd = require('./lib/addEnd');
exports.removeStart = require('./lib/removeStart');
exports.removeEnd = require('./lib/removeEnd');
exports.removeChars = require('./lib/removeChars');

exports.toCharArray = require('./lib/toCharArray');
exports.lines = require('./lib/lines');
exports.truncate = require('./lib/truncate');
exports.slugify = require('./lib/slugify');
exports.repeat = require('./lib/repeat');

exports.format = require('./lib/format');
exports.interpolate = require('./lib/interpolate');
exports.template = require('./lib/template');

exports.capitalize = require('./lib/capitalize');
exports.decapitalize = require('./lib/decapitalize');
exports.camelize = require('./lib/camelize');
exports.dasherize = require('./lib/dasherize');
exports.classify = require('./lib/classify');

exports.randomString = require('./lib/randomString');
exports.randomNumber = require('./lib/randomNumber');
exports.randomHex = require('./lib/randomHex');
exports.randomBase62 = require('./lib/randomBase62');

exports.escapeHTML = require('./lib/escapeHTML');
exports.unescapeHTML = require('./lib/unescapeHTML');

exports.hashCode = require('./lib/hashCode');
exports.toString = require('./lib/toString');
